### PR TITLE
[SPARK-42881][SQL] Codegen Support for get_json_object

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The PR adds Codegen Support for get_json_object.

### Why are the changes needed?
Improve codegen coverage and performance.
Github benchmark data(https://github.com/panbingkun/spark/actions/runs/4497396473/jobs/7912952710):
image

Local benchmark data:
image

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Add new UT.
Pass GA.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 1]
```diff
@@ -140,18 +135,92 @@ case class GetJsonObject(json: Expression, path: Expression)
...
+         |} else {
+         |  $refEvaluator.setPath(${pathEval.value});
+         |}
+         |""".stripMargin
+
+    val resultType = CodeGenerator.boxedType(dataType)
+    val resultTerm = ctx.freshName("result")
+    ev.copy(code =
+      code"""
+         |${jsonEval.code + "\n"}
```
Comment (LuciferYang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:173]: I don't think we need these  `+ "\n"`

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:173]: The generated code looks like this:
image

**Comment by panbingkun** [comment]
cc @cloud-fan @wangyum @LuciferYang 
I run benchmark - org.apache.spark.sql.execution.datasources.json.JsonBenchmark, result as follow:

- CodeGen for get_json_object
https://github.com/panbingkun/spark/actions/runs/4489492515/jobs/7895384637
image

- No CodeGen for get_json_object
https://github.com/panbingkun/spark/actions/runs/4489490118/jobs/7895379568
image

**Comment by LuciferYang** [comment]
hmm... I think we should refactor `JsonBenchmark` to make get_json_object run w/ and w/o in one

**Comment by panbingkun** [comment]
> hmm... I think we should refactor `JsonBenchmark` to make get_json_object run w/ and w/o code gen in one

Ok, Let me do it.

**Comment by LuciferYang** [comment]
@panbingkun I think we should also update `JsonBenchmark-jdk11-results.txt`, `JsonBenchmark-jdk17-results.txt` and `JsonBenchmark-results.txt` in this pr due to `JsonBenchmark` updated

**Comment by panbingkun** [comment]
> @panbingkun I think we should also update `JsonBenchmark-jdk11-results.txt`, `JsonBenchmark-jdk17-results.txt` and `JsonBenchmark-results.txt` in this pr due to `JsonBenchmark` updated

Done

**Review by MaxGekk — COMMENTED** [review]
@panbingkun Could you resolve conflicts, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you resolve conflicts, please.

Let me update the results of `JsonBenchmark` again. Waiting for it.
Thank you for your review! @MaxGekk

File: sql/core/benchmarks/JsonBenchmark-jdk11-results.txt [hunk 2]
```diff
@@ -4,120 +4,121 @@ Benchmark for performance of JSON parsing
...
-Text read                                           101            103           2          9.9         100.5       1.0X
-from_json                                          1960           1965           6          0.5        1960.1       0.1X
-json_tuple                                         2226           2235          13          0.4        2226.3       0.0X
-get_json_object                                    2077           2088          12          0.5        2077.0       0.0X
+Text read                                            99            109           9         10.1          98.6       1.0X
+from_json                                          2766           2816          46          0.4        2766.0       0.0X
+json_tuple                                         3064           3077          11          0.3        3063.8       0.0X
+get_json_object wholestage off                     2897           2917          32          0.3        2897.3       0.0X
+get_json_object wholestage on                      2832           2853          22          0.4        2831.6       0.0X
```
Comment (panbingkun) [inline sql/core/benchmarks/JsonBenchmark-jdk11-results.txt:64]: Update here

File: sql/core/benchmarks/JsonBenchmark-jdk17-results.txt [hunk 3]
```diff
@@ -7,117 +7,118 @@ OpenJDK 64-Bit Server VM 17.0.7+7 on Linux 5.15.0-1037-azure
...
-Text read                                            95            100           9         10.5          95.1       1.0X
-from_json                                          1638           1646           7          0.6        1638.5       0.1X
-json_tuple                                         1971           1996          39          0.5        1970.6       0.0X
-get_json_object                                    1799           1809          13          0.6        1799.3       0.1X
+Text read                                            90             91           3         11.2          89.5       1.0X
+from_json                                          2048           2057          12          0.5        2047.7       0.0X
+json_tuple                                         2334           2340           6          0.4        2334.1       0.0X
+get_json_object wholestage off                     2295           2299           4          0.4        2295.4       0.0X
+get_json_object wholestage on                      2158           2161           3          0.5        2158.3       0.0X
```
Comment (panbingkun) [inline sql/core/benchmarks/JsonBenchmark-jdk17-results.txt:64]: ditto

File: sql/core/benchmarks/JsonBenchmark-results.txt [hunk 4]
```diff
@@ -4,120 +4,121 @@ Benchmark for performance of JSON parsing
...
-Text read                                           165            170           4          6.1         164.7       1.0X
-from_json                                          2339           2386          77          0.4        2338.9       0.1X
-json_tuple                                         2667           2730          55          0.4        2667.3       0.1X
-get_json_object                                    2627           2659          32          0.4        2627.1       0.1X
+Text read                                           149            152           4          6.7         148.9       1.0X
+from_json                                          2103           2124          21          0.5        2103.4       0.1X
+json_tuple                                         2482           2490           7          0.4        2481.7       0.1X
+get_json_object wholestage off                     2241           2249          12          0.4        2240.7       0.1X
+get_json_object wholestage on                      2124           2132           8          0.5        2123.9       0.1X
```
Comment (panbingkun) [inline sql/core/benchmarks/JsonBenchmark-results.txt:64]: ditto

**Comment by panbingkun** [comment]
> @panbingkun Could you resolve conflicts, please.

This is done.

**Review by MaxGekk — COMMENTED** [review]
I am not sure why this should improve performance but in general LGTM. @HyukjinKwon WDYT?

**Comment by panbingkun** [comment]
> I am not sure why this should improve performance but in general LGTM. @HyukjinKwon WDYT?

In our production environment, it has indeed improved query performance for some scenarios.

**Comment by xy2953396112** [comment]
@panbingkun How much the performance improvement in your production environment？

**Comment by HyukjinKwon** [comment]
I'm fine w/ this but I don't feel confident enough to approve. I remember @viirya mentioned sth like if codegen is not implemented, the wholestage codegen chain is cut out, and it affects performance.

**Comment by LuciferYang** [comment]
> I remember @viirya mentioned sth like if codegen is not implemented, the wholestage codegen chain is cut out, and it affects performance.

This might be a valid standpoint that can support the merging of this pull request, although the benefits may not be obvious on their own.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by panbingkun** [comment]
> @panbingkun Please, resolve conflicts.

Done.

**Comment by panbingkun** [comment]
Its principle is similar to the following diagram (Although the diagram says Hive UDF Codgen)
image

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 5]
```diff
@@ -140,18 +135,92 @@ case class GetJsonObject(json: Expression, path: Expression)
...
 
   override def eval(input: InternalRow): Any = {
-    val jsonStr = json.eval(input).asInstanceOf[UTF8String]
+    evaluator.setJson(json.eval(input).asInstanceOf[UTF8String])
+    evaluator.setPath(path.eval(input).asInstanceOf[UTF8String])
+    evaluator.evaluate()
+  }
+
+  protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val refEvaluator = ctx.addReferenceObj("evaluator", evaluator)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:148]: If you initialize `evaluator` in generated code, you may not need to make `GetJsonObjectEvaluator` extend `Serializable`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 6]
```diff
@@ -140,18 +135,92 @@ case class GetJsonObject(json: Expression, path: Expression)
...
+}
+
+class GetJsonObjectEvaluator(path: Expression) extends Serializable {
+  import com.fasterxml.jackson.core.JsonToken._
+  import PathInstruction._
+  import SharedFactory._
+  import WriteStyle._
+
+  @transient
+  private lazy val parsedPath = parsePath(path.eval().asInstanceOf[UTF8String])
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:199]: Hmm, this looks weird as in codegen you will call both interpreted and codgen of `path`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 5 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 6 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Let's put this logic `path.eval().asInstanceOf[UTF8String]` outside.

**Comment by panbingkun** [comment]
Friendly ping @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 7]
```diff
@@ -140,18 +135,114 @@ case class GetJsonObject(json: Expression, path: Expression)
...
+      s"""
+         |if (${pathEval.isNull}) {
+         |  $evaluator.setPath(null);
+         |} else {
+         |  $evaluator.setPath(${pathEval.value});
+         |}
+         |""".stripMargin
+    } else {
+      ""
+    }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:186]: If path is foldable but evaluated value is null, don't you need to set a null path to `$evaluator`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:186]: Oh, nvm, I saw the default constructor sets it to null.

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 8]
```diff
@@ -1392,4 +1393,25 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+      val plan = df.queryExecution.executedPlan
+      assert(plan.isInstanceOf[WholeStageCodegenExec])
+      checkAnswer(df, Seq(Row("1", "value1")))
+    }
+  }
+
+  test("path is null") {
+    val df: DataFrame = Seq(("""{"name": "alice", "age": 5}""", "")).toDF("a", "b")
+    checkAnswer(df.selectExpr("get_json_object(a, null)"), Row(null))
+  }
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:1411]: Could you also check if this is wholestage codegen?

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 9]
```diff
@@ -1392,4 +1393,25 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
...
+
+  test("json is null") {
+     val df: DataFrame = Seq(("""{"name": "alice", "age": 5}""", "")).toDF("a", "b")
+    checkAnswer(df.selectExpr("get_json_object(null, '$.name')"), Row(null))
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:1415]: ditto.

**Review by viirya — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala (hunk 9 above)
Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: All Done.

**Comment by MaxGekk** [comment]
@panbingkun Could you rebase this PR on the recent master and resolve conflicts, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you rebase this PR on the recent master and resolve conflicts, please.

@MaxGekk Done, Thanks.

**Comment by MaxGekk** [comment]
@panbingkun Could you rebase this on recent master, please.

**Comment by panbingkun** [comment]
> @panbingkun Could you rebase this on the recent master, please.

Okay.

**Comment by MaxGekk** [comment]
The failed tests like this:
```
python/pyspark/sql/tests/pandas/test_pandas_map.py.test_other_than_dataframe_iter
[Errno 111] Connection refused
```
has been passed already a couple commits ago. Since that only the PR only rebased.

+1, LGTM. Merging to master.
Thank you, @panbingkun and @viirya @LuciferYang for review.

File: sql/core/benchmarks/JsonBenchmark-results.txt [hunk 10]
```diff
@@ -3,127 +3,128 @@ Benchmark for performance of JSON parsing
 ================================================================================================
 
 Preparing data for benchmarking ...
-OpenJDK 64-Bit Server VM 17.0.8+7-LTS on Linux 5.15.0-1046-azure
+OpenJDK 64-Bit Server VM 17.0.8+7-LTS on Linux 5.15.0-1047-azure
```
Comment (LuciferYang) [inline sql/core/benchmarks/JsonBenchmark-results.txt:6]: Sorry for the late reply, but I have to say, we also need to update `JsonBenchmark-jdk21-results.txt` ... @panbingkun

Comment (panbingkun) [reply sql/core/benchmarks/JsonBenchmark-results.txt:6]: Okay,I will sumbit a followup pr to complete it

Comment (panbingkun) [reply sql/core/benchmarks/JsonBenchmark-results.txt:6]: Followup pr: https://github.com/apache/spark/pull/43346

**Pull request closed without merging**